### PR TITLE
Avoid useless memory copies in IPC reader.

### LINKED ISSUE #2437: Avoid unecessary copies in Arrow IPC reader
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
The Arrow IPC format is designed to avoid memory copies when moving data from one implementation to another. However, as noted by @tustvold  on https://github.com/apache/arrow-rs/pull/2369#discussion_r944214515, the arrow-rs ipc reader implementation is actually copying data unnecessarily

**Describe the solution you'd like**
In the ipc code, create a Buffer initially and rewrite the ipc implementation in terms of `Buffer` rather than `&[u8]`  (as the final output needs to be in a `Buffer`

**Describe alternatives you've considered**
N/A

**Additional context**

Came up in the context of https://github.com/apache/arrow-rs/pull/2369

Possibly also related to https://github.com/apache/arrow-rs/issues/189

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>

# Which issue does this PR close?

Closes #2437.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?
Yes.

### CODE REVIEW

File: arrow/src/ipc/reader.rs [hunk 1]
```diff
@@ -53,14 +53,11 @@ fn read_buffer(
 ) -> Result<Buffer> {
     let start_offset = buf.offset() as usize;
     let end_offset = start_offset + buf.length() as usize;
-    let buf_data = &a_data[start_offset..end_offset];
+    let buf_data = Buffer::from(&a_data[start_offset..end_offset]);
```
Comment (HaoYang670) [inline arrow/src/ipc/reader.rs:56]: As we cannot slice on Buffer with length, we cannot implement `a_data` as a `Buffer`

Comment (tustvold) [reply arrow/src/ipc/reader.rs:56]: Looking at the calculation above, I think we don't need to truncate the length, as end_offset is the end of the buffer?

Comment (HaoYang670) [reply arrow/src/ipc/reader.rs:56]: The `end_offset` is the end of `buf`, but not sure whether it is the end of `a_data`

Comment (tustvold) [reply arrow/src/ipc/reader.rs:56]: Aah yes, brain fart on my part. R.e. making a_data a Buffer, we could always pass the length separately...

**Review by tustvold — APPROVED** [review]
There is definitely more that could be done here, but this is a solid start :+1:

**Comment by HaoYang670** [comment]
Thank you, @tustvold 
I just have a question (maybe it is an old question). Why don't we support truncate by length for `Buffer`.

**Comment by tustvold** [comment]
I don't believe there is a particular reason, I think lifting length out of `Bytes` into `Buffer` would be a sensible change imo

**Comment by alamb** [comment]
Wow that was fast 🚀

File: arrow/src/ipc/compression/codec.rs [hunk 2]
```diff
@@ -92,18 +92,16 @@ impl CompressionCodec {
...
         let buffer = if decompressed_length == 0 {
             // emtpy
-            let empty = Vec::<u8>::new();
-            Buffer::from(empty)
+            Buffer::from([])
         } else if decompressed_length == LENGTH_NO_COMPRESSED_DATA {
             // no compression
-            let data = &input[(LENGTH_OF_PREFIX_DATA as usize)..];
-            Buffer::from(data)
+            input.slice(LENGTH_OF_PREFIX_DATA as usize)
```
Comment (alamb) [inline arrow/src/ipc/compression/codec.rs:104]: ❤️

**Pull request merged**